app/services/bulk_upload_critters.rb

```diff
@@ -0,0 +1,65 @@
+require 'csv'
+class BulkUploadCritters < ApplicationService
+  def initialize(file_data:)
+    @file_data = file_data
+    @errors = []
+    @new_records = []
+    @updated_records = []
+  end
+
+  def call
+    process_csv
+    result
+  rescue => exception
+    failure(exception.message)
+  end
+
+  private
+
+  def process_csv
+    CSV.foreach(@file_data.tempfile, headers: true) do |row|
+      critter_attrs = {
+        name: row['critter name']
+      }
+      critter = Critter.find_by(id: row['id'])
+      if critter
+        update_critter!(critter: critter, attrs: critter_attrs)
+      else
+        create_critter!(attrs: critter_attrs)
+      end
+    end
+  end
+
+  def update_critter!(critter:, attrs:)
+    if critter.update(attrs)
+      @updated_records << critter
+    else
+      # TODO: improve error handling/reporting
+      @errors << critter.errors
+    end
```

Comment:
find_or_create_by does _not_ update records. i confirmed by testing it out, but also here's the [source code](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_create_by):

```ruby
# File activerecord/lib/active_record/relation.rb, line 189
    def find_or_create_by(attributes, &block)
      find_by(attributes) || create(attributes, &block)
    end
```

Reply:
I also rediscovered that today, so it's not applicable to use in this case.